### PR TITLE
Introduction of PackageWrapper & ResultWrapper

### PR DESCRIPTION
**Pull request by BvB93**
See issue https://github.com/SCM-NV/qmflows/issues/147.

Major
------
* Introduced the ``PackageWrapper`` and ``ResultWrapper`` subclasses.
  ``PackageWrapper`` will take a single [``plams.Job``](https://www.scm.com/doc/plams/components/jobs.html#job-api) type as argument and automatically return the appropiate [``Package``](https://qmflows.readthedocs.io/en/latest/packages.html) instance upon calling ``PackageWrapper.__call__()``.

```
>>> from scm.plams import ADFJob, AMSJob, Molecule
>>> from qmflows import Settings, PackageWrapper, run

>>> from qmflows.packages.SCM import ADF_Result
>>> from qmflows.packages.package_wrapper import ResultWrapper

>>> mol = Molecule(...)
>>> s1 = Settings(...)
>>> s2 = Settings(...)

>>> job_ams = PackageWrapper(AMSJob)(s1, mol, ...)
>>> job_adf = PackageWrapper(ADFJob)(s2, mol, ...)

>>> result_ams: ResultWrapper = run(job_ams, ...)
>>> result_adf: ADF_Result = run(job_adf, ...)

```

* Allow ``Package.generic2specific()`` to handle PLAMS style input settings. 
  For example, ``settings.input.blablabla`` is now automatically converted into  ``settings.specific[self.pkg_name].blablabla``.

Minor
------
* Changed the ``Package.generic_dict_file`` attribute from an instance- to a class-variable, as it's value is already constant with respect to the class type.
* Changed the default value of ``Package.generic_dict_file`` to [``NotImplemented``](https://docs.python.org/3/library/constants.html#NotImplemented). Only applies to the [``Package``](https://qmflows.readthedocs.io/en/latest/packages.html) baseclass, not its subclasses.
* Got rid of a lot of nesting in ``Package.generic2specific()``.
* Updated the Sphinx ``conf.py`` settings.
* Changed minimum Sphinx version to 2.1.
* Added documentation for the new ``PackageWrapper`` class.
* Added ``PackageWrapper`` to the main ``__init__.py`` file.
* Added new tests.
* Bug fix in ``init_restart()`` and ``InitRestart``: only attempt to delete the newly created workdir if there is an actual previously existing workdir.

### CODE REVIEW

File: src/qmflows/packages/packages.py [hunk 1]
```diff
@@ -348,38 +359,44 @@ def generic2specific(self, settings: Settings,
 
         specific_from_generic_settings = Settings()
         for k, v in settings.items():
-            if k != "specific":
```
Comment (felipeZ) [inline src/qmflows/packages/packages.py]: Thanks for improving this mess!

File: test/test_using_plams/test_package_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+from scm.plams import ADFJob, AMSJob, from_smiles
+
+from qmflows import PackageWrapper, run, Settings
+from qmflows.packages.SCM import ADF_Result
+from qmflows.packages.package_wrapper import ResultWrapper
+
+PATH = Path('test') / 'test_files'
+
+
+@pytest.mark.slow
```
Comment (felipeZ) [inline test/test_using_plams/test_package_wrapper.py]: You can also mock the ADF

Comment (BvB93) [reply test/test_using_plams/test_package_wrapper.py]: Would it be better to mock ADF or simply raise an exception when ADF cannot be found?
By checking for ADF environment variables for example?

Comment (felipeZ) [reply test/test_using_plams/test_package_wrapper.py]: Well, the point is that for CI it is difficult to install ADF, so in general is more reliable to have a mock of the ADF result

Comment (BvB93) [reply test/test_using_plams/test_package_wrapper.py]: Fair enough, the test is aready marked as ``slow`` though, just like all other tests requiring ADF.

Comment (felipeZ) [reply test/test_using_plams/test_package_wrapper.py]: If you consider that is important, you can mock the ADF result like https://github.com/SCM-NV/qmflows/blob/master/test/test_mock_operations.py

Comment (BvB93) [reply test/test_using_plams/test_package_wrapper.py]: Thanks for the info, I'll get back to you a bit later.

File: test/test_using_plams/test_package_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+from qmflows.packages.packages import Result
+from qmflows.packages.SCM import ADF_Result
+from qmflows.packages.package_wrapper import ResultWrapper
+
+PATH = Path('test') / 'test_files'
+ADF_ENVIRON = frozenset({'ADFBIN', 'ADFHOME', 'ADFRESOURCES', 'SCMLICENSE'})
+HAS_ADF: bool = ADF_ENVIRON.issubset(os.environ.keys())
+
+
+def _get_result(promised_object: PromisedObject, job: Type[Job]) -> Result:
```
Comment (BvB93) [inline test/test_using_plams/test_package_wrapper.py]: Continuation of https://github.com/SCM-NV/qmflows/pull/149#discussion_r387583432:
> Thanks for the info, I'll get back to you a bit later.

Allright, I've created a way to (sort of) mock the ADF results.
More specifically, it creates a normal-ish ``qmflows.Result`` instance and then attaches a pickled .dill file from a previous calculation.

File: test/test_using_plams/test_package_wrapper.py (hunk 2 above)
Comment (BvB93) [reply test/test_using_plams/test_package_wrapper.py]: See https://github.com/SCM-NV/qmflows/pull/149#discussion_r387616762.

**Pull request merged**